src/headless/converse-muc.js

```diff
@@ -507,6 +507,42 @@ converse.plugins.add('converse-muc', {
                     }, null, 'message', 'groupchat', null, room_jid,
                     {'matchBareFromJid': true}
                 );
+                this.muc_notifications_handler = _converse.connection.addHandler(stanza => {
+                    const item = sizzle(`x[xmlns="${Strophe.NS.MUC_USER}"]`, stanza).pop();
+
+                    if (item) {
+                        const from = stanza.getAttribute("from");
+                        const type = stanza.getAttribute("type");
+                        const affiliation = item.getAttribute('affiliation');
+                        const jid = item.getAttribute('jid');
+
+                        const data = {
+                          'from': from,
+                          'nick': Strophe.getNodeFromJid(jid),
+                          'type': type,
+                          'states': [],
+                          'show': type == 'unavailable' ? 'offline' : 'online',
+                          'affiliation': affiliation,
+                          'role': item.getAttribute('role'),
+                          'jid': jid
+                        }
+
+                        const attributes = Object.assign(data, {
+                            'jid': Strophe.getBareJidFromJid(data.jid),
+                            'resource': Strophe.getResourceFromJid(data.jid)
+                        });
+                        const occupant = this.occupants.findOccupant({'jid': data.jid});
+
+                        if (occupant) {
+                          occupant.save(attributes);
+                        } else {
+                          this.occupants.create(attributes);
+                        }
+                    }
+                    return true;
+
+                }, null, 'message', null, null, room_jid,
```

Comment:
You can specify the xmlns here (replace the first `null` with `Strophe.NS.MUC_USER`, then you don't have to check whether the item exists on line 513 and then this handler will only fire when a child of the stanza has that xmlns.

Reply:
We might want to handle other &lt;message&gt; child stanzas with different namespaces here in future. Don't you think?

Reply:
No, probably not, and even if that were the case, then we can simply loosen the restriction again.

The way it currently is, this handler gets called unnecessarily probably 99% of the time.

Reply:
I still need to extract the child element from the stanza, so the test is still required

Reply:
> we can simply loosen the restriction again.

ok